async-iterator.md

```diff
@@ -0,0 +1,904 @@
+
+I'll share some notes about the design for async-iterator methods (C# 8.0 feature).
+
+Editorial notes:
+- Some sections probably should be skipped on first reading. I marked them as "optional".
+- Although I will gloss over many important parts, I will try to mention them to give a sense of how things fit together.
+- `/* ... */` comments indicate omitted code, while `// ... ` comments apply to code that is there.
+- Every illustration of generated code omits some details.
+
+
+## Overview
+
+Async-iterator methods are methods with both yields and awaits and produce async-enumerables (`IAsyncEnumerable<T>`) or async-enumerators (`IAsyncEnumerator<T>`).
+Those are methods that yield a sequence of items (like iterator methods that return `IEnumerable<T>`), but allow for asynchronous computation (like async methods).
+
+
+## Consumption with `await foreach`
+
+The `await foreach` statement offers the most straightforward way of enumerating an async-enumerable's items.
+It is very similar to `foreach` for enumerables. The main difference is that `await foreach` awaits each item (and also disposal).
+
+The parallel jumps out in the interfaces involved:
+
+| async enumerables | enumerables |
+| --- | --- |
+| `IAsyncEnumerable<T>` used with `GetAsyncEnumerator()` | `IEnumerable<T>` used with `GetEnumerator()` |
+| `IAsyncEnumerator<T>` used with `await MoveNextAsync()` and `Current` | `IEnumerator<T>` used with `MoveNext()` and `Current` |
```

Comment:
interesting.  so we moved away from the API choice where there would be one expensive async bulk-retrieval, followed by an inexpensive enumeration of the retrieved elements?